File: frontend/src/lib/crypto/graine.ts
/**
 * Graine: seed to message key, and the AES-GCM seal built on it.
 *
 * This module is deliberately PURE - no storage, no network, no clock, no singletons. It is the
 * piece that has to be provably right in isolation, so everything it needs arrives as an argument
 * and everything it produces is a return value. Session lifetime, rotation and distribution live
 * above it.
 *
 * Protocol, measurements and the alternatives it rules out:
 * `docs/wiki/protocols/channel-encryption.md`.
 */

import { fromBase64, toBase64 } from '$lib/utils/hex';
import {
  GRAINE_HKDF_INFO,
  GRAINE_MESSAGE_KEY_BYTES,
  GRAINE_NONCE_BYTES,
  GRAINE_SEED_BYTES,
} from '$lib/crypto/graineConstants';

/** Thrown when an argument cannot produce a key, before any crypto call is attempted. */
export class GraineInputError extends Error {
  constructor(message: string) {
    super(message);
    this.name = 'GraineInputError';
  }
}

/**
 * A fresh 32-byte seed. One per (sender, channel, session); never derived from anything, because
 * the only thing that must be unguessable here is the seed itself.
 */
export function newGraineSeed(): Uint8Array {
  return crypto.getRandomValues(new Uint8Array(GRAINE_SEED_BYTES));
}

/**
 * A fresh session id: 18 random bytes, base64url.
 *
 * Random rather than a counter, and therefore needing no coordination: **no two senders ever write
 * the same session namespace**, so there is nothing here for two devices to race over and no
 * election to hold. base64url so it is safe in a JSON key, a file name and a push payload without
 * escaping - which the native mirror relies on, since it reads this as a plain map key.
 */
export function newGraineSessionId(): string {
  const raw = crypto.getRandomValues(new Uint8Array(18));
  return toBase64(raw).replace(/\+/g, '-').replace(/\//g, '_').replace(/=+$/, '');
}

/**
 * The key for message `index` of a session.
 *
 * **Derived from the seed rather than ratcheted forward, and that is a deliberate difference from
 * megolm.** A forward ratchet only computes index i from index i-1, so reading an arbitrary
 * message means either replaying from zero or keeping every intermediate state. Channel history is
 * served by REST, newest first, in whatever order the reader scrolls - so an arbitrary index has to
 * be cheap, and it is exactly the case a forward ratchet is worst at. What the ratchet would have
 * bought (a key recovered at index i does not open i-1) is bought here too: each index is an
 * independent HKDF output, so one message key opens exactly one message. Only the SEED opens the
 * session, which is the thing rotation bounds.
 *
 * The derivation binds the session id, so a seed replayed under a different session id yields
 * different keys and its messages simply fail to open. The id travels as the HKDF salt and the
 * index as fixed-width big-endian bytes in the info, which leaves no way for two different
 * (id, index) pairs to produce the same input - the ambiguity a concatenated string would have.
 */
export async function deriveMessageKey(
  seed: Uint8Array,
  sessionId: string,
  index: number
): Promise<CryptoKey> {
  const bits = await deriveMessageKeyBytes(seed, sessionId, index);
  return crypto.subtle.importKey(
    'raw',
    bits as BufferSource,
    { name: 'AES-GCM', length: GRAINE_MESSAGE_KEY_BYTES * 8 },
    false,
    ['encrypt', 'decrypt']
  );
}

/**
 * The raw bytes {@link deriveMessageKey} imports, exported for exactly two reasons and no others.
 *
 * A `CryptoKey` is non-extractable by design, which is right for every caller - and which would
 * make this derivation unpinnable by a test vector, so a reword of the info string or a change of
 * endianness would pass every test while silently breaking every other implementation. The native
 * push decoder is the other implementation: it re-derives these bytes in Rust before any WebView
 * runs, and the vectors in `graine.test.ts` are the contract the two sides are held to.
 *
 * Message code must call {@link deriveMessageKey} and never this.
 */
export async function deriveMessageKeyBytes(
  seed: Uint8Array,
  sessionId: string,
  index: number
): Promise<Uint8Array> {
  if (seed.length !== GRAINE_SEED_BYTES) {
    throw new GraineInputError(`Seed must be ${GRAINE_SEED_BYTES} bytes, got ${seed.length}`);
  }
  if (!sessionId) {
    throw new GraineInputError('Session id is required: it is bound into the derivation');
  }
  if (!Number.isInteger(index) || index < 0 || index > 0xffffffff) {
    throw new GraineInputError(`Message index must be a uint32, got ${index}`);
  }

  const material = await crypto.subtle.importKey('raw', seed as BufferSource, 'HKDF', false, [
    'deriveBits',
  ]);
  const bits = await crypto.subtle.deriveBits(
    {
      name: 'HKDF',
      hash: 'SHA-256',
      salt: new TextEncoder().encode(sessionId) as BufferSource,
      info: messageKeyInfo(index) as BufferSource,
    },
    material,
    GRAINE_MESSAGE_KEY_BYTES * 8
  );
  return new Uint8Array(bits);
}

/** `"canari-graine-v1"` followed by the index as 4 big-endian bytes. Fixed width, so unambiguous. */
function messageKeyInfo(index: number): Uint8Array {
  const label = new TextEncoder().encode(GRAINE_HKDF_INFO);
  const info = new Uint8Array(label.length + 4);
  info.set(label, 0);
  new DataView(info.buffer).setUint32(label.length, index, false);
  return info;
}

/** A sealed message: what goes on the wire, base64 for a JSON body. */
export interface GraineSealed {
  ciphertext: string;
  nonce: string;
}

/**
 * Seals `plaintext` for message `index` of a session.
 *
 * A fresh random nonce per message rather than one derived from the index: the index is public and
 * a derived nonce would repeat the moment a seed were ever reused across sessions - the one
 * catastrophic failure mode of GCM. Twelve random bytes cost twelve bytes.
 */
export async function sealWithGraine(
  seed: Uint8Array,
  sessionId: string,
  index: number,
  plaintext: Uint8Array
): Promise<GraineSealed> {
  const key = await deriveMessageKey(seed, sessionId, index);
  const nonce = crypto.getRandomValues(new Uint8Array(GRAINE_NONCE_BYTES));
  const sealed = await crypto.subtle.encrypt(
    { name: 'AES-GCM', iv: nonce as BufferSource },
    key,
    plaintext as BufferSource
  );
  return { ciphertext: toBase64(new Uint8Array(sealed)), nonce: toBase64(nonce) };
}

/**
 * Opens a message sealed by {@link sealWithGraine}.
 *
 * Rejects rather than returning null: a failure here means the seed, the session id or the index is
 * wrong, or the ciphertext was tampered with, and a caller that treated any of those as "no
 * message" would render an unreadable channel as an empty one - two different facts.
 */
export async function openWithGraine(
  seed: Uint8Array,
  sessionId: string,
  index: number,
  sealed: GraineSealed
): Promise<Uint8Array> {
  const key = await deriveMessageKey(seed, sessionId, index);
  const plaintext = await crypto.subtle.decrypt(
    { name: 'AES-GCM', iv: fromBase64(sealed.nonce) as BufferSource },
    key,
    fromBase64(sealed.ciphertext) as BufferSource
  );
  return new Uint8Array(plaintext);
}

File: frontend/src-tauri/src/mobile/graine.rs
//! Graine message-key derivation, native half.
//!
//! A community-channel push is decrypted before any WebView runs, so the derivation cannot be
//! called into the browser - it has to exist twice. This is the ONLY duplicate of
//! `frontend/src/lib/crypto/graine.ts`, it is deliberate, and the two are held together by the
//! same four test vectors: they are asserted in `graine.test.ts` on the TS side and in the tests at
//! the bottom of this file, both produced independently from the specification below.
//!
//! `key(index) = HKDF-SHA256(ikm = seed, salt = utf8(session_id), info = LABEL || be32(index))`
//!
//! Changing LABEL, the salt, or the width of the index is a protocol change and must land on both
//! sides in one commit, or a phone stops being able to read its own notifications.
//!
//! Protocol: `docs/wiki/protocols/channel-encryption.md`.

use hkdf::Hkdf;
use sha2::Sha256;

/// HKDF `info` prefix and the wire version. Mirrors `GRAINE_HKDF_INFO` in `graineConstants.ts`.
pub const GRAINE_HKDF_INFO: &[u8] = b"canari-graine-v1";

/// Seed length, and therefore the length of every key derived from it (AES-256).
pub const GRAINE_SEED_BYTES: usize = 32;

/// Why a derivation could not happen, as a type rather than a sentence.
#[derive(Debug, PartialEq, Eq)]
pub enum GraineError {
    /// The seed was not exactly [`GRAINE_SEED_BYTES`] long.
    SeedLength(usize),
    /// The session id was empty - it is bound into the derivation and cannot be defaulted.
    MissingSessionId,
}

/// The AES-256-GCM key for message `index` of `session_id`.
///
/// Derived from the seed rather than ratcheted forward, matching the TS side and for the same
/// reason: channel history is served newest-first over REST, so an arbitrary index has to be cheap
/// to reach - the one case a forward ratchet is worst at. Each index is an independent HKDF output,
/// so one message key opens exactly one message and only the seed opens the session.
pub fn derive_message_key(
    seed: &[u8],
    session_id: &str,
    index: u32,
) -> Result<[u8; 32], GraineError> {
    if seed.len() != GRAINE_SEED_BYTES {
        return Err(GraineError::SeedLength(seed.len()));
    }
    if session_id.is_empty() {
        return Err(GraineError::MissingSessionId);
    }

    let mut info = Vec::with_capacity(GRAINE_HKDF_INFO.len() + 4);
    info.extend_from_slice(GRAINE_HKDF_INFO);
    info.extend_from_slice(&index.to_be_bytes());

    let hk = Hkdf::<Sha256>::new(Some(session_id.as_bytes()), seed);
    let mut out = [0u8; 32];
    // Only fails for an output longer than 255 * HashLen, which 32 bytes cannot be.
    hk.expand(&info, &mut out)
        .expect("32 bytes is always a valid HKDF output length");
    Ok(out)
}

#[cfg(test)]
mod tests {
    use super::*;
    use base64::{engine::general_purpose::STANDARD, Engine as _};

    /// Bytes 0..31, the same seed the TS vectors use.
    fn seed() -> [u8; 32] {
        let mut s = [0u8; 32];
        for (i, b) in s.iter_mut().enumerate() {
            *b = i as u8;
        }
        s
    }

    /// The contract between this file and `frontend/src/lib/crypto/graine.test.ts`.
    ///
    /// Both suites assert the SAME four values, produced independently of either implementation.
    /// A reworded label, a flipped endianness or a dropped session id passes a round-trip test on
    /// each side separately and breaks the pair - which is the failure these vectors exist to catch.
    #[test]
    fn matches_the_shared_test_vectors() {
        let cases: [(&str, u32, &str); 4] = [
            (
                "test-session",
                0,
                "go/kAahuKrCvwB4CquM4zrgRTQyH2+WlHLNQjSB2VKA=",
            ),
            (
                "test-session",
                7,
                "Pnqk2gdQBRQe4p2hXM+vPGNQwHh0Ukc8tcRTnq1fcNw=",
            ),
            (
                "other-session",
                0,
                "n5USTdC3HUsFjhOOwUcw23nwoWgJbl7A6f/EJOToaFQ=",
            ),
            (
                "test-session",
                u32::MAX,
                "f/0pgQZqn8yiwLK3UiG2D6O1RrIqFp+BMeonMygzsIs=",
            ),
        ];
        for (session_id, index, expected) in cases {
            let key = derive_message_key(&seed(), session_id, index).expect("derives");
            assert_eq!(STANDARD.encode(key), expected, "{session_id} @ {index}");
        }
    }

    #[test]
    fn every_index_gets_its_own_key() {
        let a = derive_message_key(&seed(), "s", 0).unwrap();
        let b = derive_message_key(&seed(), "s", 1).unwrap();
        assert_ne!(a, b);
    }

    #[test]
    fn refuses_input_that_cannot_produce_a_key() {
        assert_eq!(
            derive_message_key(&[0u8; 16], "s", 0),
            Err(GraineError::SeedLength(16))
        );
        assert_eq!(
            derive_message_key(&seed(), "", 0),
            Err(GraineError::MissingSessionId)
        );
    }
}
